### PR TITLE
feat: HP constraints harness exception handling [DET-4867]

### PR DESCRIPTION
**Pull request by naren-determined**
## Description

Enable the harness to catch `det.InvalidHP` exceptions raised in user code, and propagate that information to the searcher. Note: this is a user-facing API change, thought not a breaking change.

## Test Plan

In any `model_def.py` raise a `det.InvalidHP("msg")` in either `__init__`, `train_batch`, or `evaluate_batch` for PyTorch models. For Keras and Estimator models, raise in either `__init__` or in a callback with validation metrics. Ensure the trial stops training, and that an InvalidHP is raised (should be apparent in the trial logs).

## Checklist

- [x] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: harness/determined/layers/_workload_manager.py [hunk 1]
```diff
@@ -132,7 +132,7 @@ def yield_train_for_step(
             )
 
         def _respond(in_response: workload.Response) -> None:
-
+            # raise ValueError(in_response)
```
Comment (shiyuann) [inline harness/determined/layers/_workload_manager.py:135]: nit: remove this line

File: harness/determined/_train_context.py [hunk 2]
```diff
@@ -144,6 +145,26 @@ def set_stop_requested(self, stop_requested: bool) -> None:
         )
         self._stop_requested = stop_requested
 
+    def get_invalid_hp(self) -> bool:
```
Comment (shiyuann) [inline harness/determined/_train_context.py:148]: Is this intended to be used by users?

File: harness/determined/_train_context.py [hunk 3]
```diff
@@ -144,6 +145,26 @@ def set_stop_requested(self, stop_requested: bool) -> None:
...
+        """
+        Return whether a trial invalid hp has been requested.
+        """
+        return self._invalid_hp
+
+    def set_invalid_hp(self, invalid_hp: bool) -> None:
```
Comment (shiyuann) [inline harness/determined/_train_context.py:154]: Is this supposed to be used by users or should they just raise exception?

File: harness/determined/exec/harness.py [hunk 4]
```diff
@@ -129,21 +130,22 @@ def build_and_run_training_pipeline(env: det.EnvContext) -> None:
...
 
                 with det._catch_sys_exit():
-                    controller = load.prepare_controller(
-                        env,
-                        iter(workload_mgr),
-                        load_path,
-                        socket_mgr.get_rendezvous_info(),
-                        hvd_config,
-                    )
+                    with det._catch_init_invalid_hp(workloads):
```
Comment (shiyuann) [inline harness/determined/exec/harness.py:141]: If you catch the exception here, you will only be able to catch when executing users' trial init function.

File: harness/determined/exec/harness.py [hunk 5]
```diff
@@ -224,4 +226,8 @@ def main() -> None:
 
 
 if __name__ == "__main__":
-    main()
+    try:
```
Comment (shiyuann) [inline harness/determined/exec/harness.py:229]: blocking: you should actually catch the exception on running `prepare_controller()` and `controller.run()` not at the top level.

File: harness/determined/pytorch/_pytorch_trial.py [hunk 6]
```diff
@@ -143,18 +143,51 @@ def _set_data_loaders(self) -> None:
...
-                )
+                except det.InvalidHP as e:
+                    logging.info(
+                        "Invalid hyperparameter exception in trial validation step: {}".format(e)
+                    )
+                    response_func(
+                        util.wrap_metrics(
+                            self._compute_validation_metrics(),
+                            self.context.get_stop_requested(),
+                            self.context.set_invalid_hp(True),
```
Comment (shiyuann) [inline harness/determined/pytorch/_pytorch_trial.py:187]: I think you actually intended to just put a `True` here rather than `self.context.set_invalid_hp(True),`

File: harness/determined/pytorch/_pytorch_trial.py [hunk 7]
```diff
@@ -143,18 +143,51 @@ def _set_data_loaders(self) -> None:
...
+                        )
+                    )
+                    raise
             elif w.kind == workload.Workload.Kind.COMPUTE_VALIDATION_METRICS:
-                response_func(
-                    util.wrap_metrics(
-                        self._compute_validation_metrics(), self.context.get_stop_requested()
+                try:
```
Comment (shiyuann) [inline harness/determined/pytorch/_pytorch_trial.py:171]: I think you actually want to put only `self._compute_validation_metrics()` in the "try" rather than the call to `response_func`.

File: harness/determined/pytorch/_pytorch_trial.py [hunk 8]
```diff
@@ -143,18 +143,51 @@ def _set_data_loaders(self) -> None:
     def run(self) -> None:
         for w, args, response_func in self.workloads:
             if w.kind == workload.Workload.Kind.RUN_STEP:
-                response_func(
-                    util.wrap_metrics(
-                        self._train_for_step(w.step_id, w.num_batches, w.total_batches_processed),
-                        self.context.get_stop_requested(),
+                try:
+                    response_func(
```
Comment (shiyuann) [inline harness/determined/pytorch/_pytorch_trial.py:147]: same

File: harness/determined/keras/_tf_keras_trial.py [hunk 9]
```diff
@@ -773,19 +773,41 @@ def _control_loop(self) -> None:
...
+                        "Invalid hyperparameter exception in trial train step: {}".format(e)
+                    )
+                    response_func(util.wrap_metrics({}, False, True))
+                    raise
             elif wkld.kind == workload.Workload.Kind.COMPUTE_VALIDATION_METRICS:
-                response_func(
-                    det.util.wrap_metrics(
-                        self._compute_validation_metrics(), self.context.get_stop_requested()
+                try:
+                    response_func(
```
Comment (shiyuann) [inline harness/determined/keras/_tf_keras_trial.py:792]: same

File: harness/determined/keras/_tf_keras_trial.py [hunk 10]
```diff
@@ -773,19 +773,41 @@ def _control_loop(self) -> None:
...
             logging.debug(f"Received wkld {wkld.kind} with args {args}.")
             if wkld.kind == workload.Workload.Kind.RUN_STEP:
-                # Configure the state for a training step.
-                self.train_response_func = response_func
-                self.train_workload_batches = 0
-                self.train_workload_metrics = []
-                self.train_workload_len = wkld.num_batches
-                self.multiplexer.set_batches_requested(wkld.num_batches)
-                break
+                try:
```
Comment (shiyuann) [inline harness/determined/keras/_tf_keras_trial.py:776]: There is no need to capture an exception here. You should instead capture the exception in `TFKerasTrialController.run`.

**Review by shiyuann — COMMENTED** [review]
@rb-determined-ai probably has an opinion on how this should be written in the harness. One question I have is: have we already merged the change in the master? If not, this might break the master if you haven't confirmed.

File: harness/determined/exec/harness.py (hunk 4 above)
Comment (naren-determined) [reply harness/determined/exec/harness.py:141]: This is the intended functionality for this context manager. We will handle invalid hp exceptions raised in training and validations workloads in the individual trial controllers.

File: harness/determined/_execution.py [hunk 11]
```diff
@@ -6,10 +6,25 @@
...
+    """
+
+    def __init__(self, message: str = "Invalid hyperparameter exception encountered") -> None:
+        self.message = message
+        super().__init__(self.message)
+
+    def __str__(self) -> str:
+        return repr(self.message)
+
+
```
Comment (rb-determined-ai) [inline harness/determined/_execution.py:27]: I actually think that this would probably suffice:
```python
class InvalidHP(Exception):
    def __init__(self, msg: str = "...") -> None:
        if not isinstance(msg, str):
            raise TypeError("InvalidHP exceptions can be initialized with a custom message, "
                f"but it must be a string, not {type(msg).__name__}")
        self.msg = msg
```
I don't think the `__str__` is needed or the super().__init__.

File: harness/determined/pytorch/_pytorch_trial.py (hunk 7 above)
Comment (rb-determined-ai) [reply harness/determined/pytorch/_pytorch_trial.py:170]: I think since the `except` clause only catches InvalidHP, and since InvalidHP is only thrown by user code, and since the response_func call is quite different in the two cases, I think this is actually a pretty reasonable strategy.

File: master/pkg/searcher/asha.go [hunk 12]
```diff
@@ -272,6 +272,8 @@ func (s *asyncHalvingSearch) trialExitedEarly(
 		s.InvalidTrials++
 		// Remove metrics associated with InvalidHP trial across all rungs
 		highestRungIndex := s.TrialRungs[requestID]
+		rung := s.Rungs[highestRungIndex]
```
Comment (rb-determined-ai) [inline master/pkg/searcher/asha.go:275]: Why is this in this PR?  I don't understand how it is related.

File: harness/determined/keras/_tf_keras_trial.py [hunk 13]
```diff
@@ -650,6 +650,16 @@ def run(self) -> None:
...
+        except det.InvalidHP as e:
+            for _, _, response_func in self.workloads:
+                logging.info("Invalid hyperparameter exception in trial: {}".format(e))
+                response_func(
+                    det.util.wrap_metrics(
+                        {},
+                        self.context.get_stop_requested(),
+                        True,
+                    )
+                )
```
Comment (rb-determined-ai) [inline harness/determined/keras/_tf_keras_trial.py:662]: Hm... this is maybe not going to do quite what you are thinking.  this hack of reading from the workloads is only valid before any workloads have been read (like when you initialize the trial), but I don't see how you can guarantee that if you are catching InvalidHP exceptions from the entire fit loop.

File: harness/determined/exec/harness.py (hunk 4 above)
Comment (rb-determined-ai) [reply harness/determined/exec/harness.py:141]: It might be worth mentioning why it is designed this way (I helped with this design haha).  This is ultimately a way for us to hack around the layers abstraction of the harness, which in some cases needs to say alive long enough for the master to ask for a checkpoint (right?).

But if there's a failure at init time, we don't need to do that, and we can just exit out of the harness.

(if we could push checkpoints, then we could always decide if we want a checkpoint at the moment we catch the error, then push a checkpoint, then exit, and things would be a bit simpler)

File: harness/determined/layers/_workload_manager.py [hunk 14]
```diff
@@ -144,32 +143,42 @@ def _respond(in_response: workload.Response) -> None:
...
+                out_response = {
+                    "type": "WORKLOAD_COMPLETED",
+                    "workload": wkld,
+                    "start_time": start_time,
+                    "end_time": _current_timestamp(),
+                    "metrics": metrics,
+                }
+                out_response["exited_reason"] = "INVALID_HP"
+                respond(out_response)
+            else:
```
Comment (rb-determined-ai) [inline harness/determined/layers/_workload_manager.py:156]: Is it valid to just `return` here, rather than add an extra layer of indentation to the whole rest of the function?

File: harness/determined/exec/harness.py (hunk 5 above)
Comment (rb-determined-ai) [reply harness/determined/exec/harness.py]: Actually, I think this is the right choice.  When we catch an exception during the `trial.__init__()`, we need to send a response to the master and just exit.  We could call "sys.exit()", but we already catch sys.exit calls and assume the user messed up.  So instead the choice was to let the InvalidHP exception continue to bubble up.

But then you don't want to show an InvalidHP stacktrace necessarily, so we drop that exception here.

File: master/pkg/searcher/asha.go (hunk 12 above)
Comment (naren-determined) [reply master/pkg/searcher/asha.go:275]: This was a bug in the searcher code where some trials weren't being closed out properly, and since it's only a two line change figured I would merge it in here.

Comment (naren-determined) [reply master/pkg/searcher/asha.go:275]: Since we are bypassing the `promoteAsync` function due to an InvalidHP, we need to decrement the oustanding trials in a rung normally done by that function, so I do that here.

File: harness/determined/layers/_workload_manager.py (hunk 14 above)
Comment (naren-determined) [reply harness/determined/layers/_workload_manager.py:156]: Yep, good call here.

File: harness/determined/keras/_tf_keras_trial.py (hunk 13 above)
Comment (naren-determined) [reply harness/determined/keras/_tf_keras_trial.py:662]: @rb-determined-ai Then I should probably take a similar approach to `_pytorch_trial.py` and catch exceptions in this area of the code: https://github.com/determined-ai/determined/blob/master/harness/determined/keras/_tf_keras_trial.py#L778 for both keras and estimators?

Comment (rb-determined-ai) [reply harness/determined/keras/_tf_keras_trial.py:662]: Hm, yes I think so.   It will be a bit trickier though, because the places where user code is called are a bit more numerous, due the the Keras Callbacks support.

File: harness/determined/exec/harness.py (hunk 4 above)
Comment (shiyuann) [reply harness/determined/exec/harness.py:141]: I have added a checkpoint API in [a recent PR](https://github.com/determined-ai/determined/pull/1968). Do you think it would help this PR to get less complicated? @rb-determined-ai

File: harness/determined/exec/harness.py (hunk 5 above)
Comment (shiyuann) [reply harness/determined/exec/harness.py]: Although it works, it looks complicated and hacky. Do you know how would this be simplified in a push-arch world?

File: harness/determined/pytorch/_pytorch_trial.py (hunk 7 above)
Comment (shiyuann) [reply harness/determined/pytorch/_pytorch_trial.py:170]: > since the response_func call is quite different in the two cases, I think this is actually a pretty reasonable strategy.

I don't understand this. We only run user code in `self._compute_validation_metrics()`, which should be the only place to throw InvalidHP. How would response_func interact with InvalidHP?

File: harness/determined/exec/harness.py (hunk 5 above)
Comment (naren-determined) [reply harness/determined/exec/harness.py]: I think I changed this actually, in a recent commit so if you take another look at the harness.py file I don't think this is an issue anymore.

File: harness/determined/pytorch/_pytorch_trial.py (hunk 7 above)
Comment (naren-determined) [reply harness/determined/pytorch/_pytorch_trial.py:170]: If you look at `_workload_manager.py`, I'm handling InvalidHP as a special case when handling the response_func, and that is how information is being propagated to the searcher to handle InvalidHP

File: harness/determined/util.py [hunk 15]
```diff
@@ -114,15 +114,19 @@ def make_metrics(num_inputs: Optional[int], batch_metrics: List[Dict[str, Any]])
...
+    Skipped if not chief.
+    """
     if isinstance(metrics, det.workload.Skipped):
         return metrics
     else:
-        return {
-            "metrics": metrics,
-            "stop_requested": stop_requested,
-        }
+        return {"metrics": metrics, "stop_requested": stop_requested, "invalid_hp": invalid_hp}
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/util.py:129]: yes, for now at least.  Maybe it will change when we have a full push architecture.

File: harness/determined/exec/harness.py (hunk 4 above)
Comment (rb-determined-ai) [reply harness/determined/exec/harness.py:141]: absolutely.

File: harness/determined/estimator/_estimator_trial.py [hunk 16]
```diff
@@ -270,19 +271,44 @@ def control_loop(self) -> None:
...
-                # Break out of the control loop so that the train process
-                # re-enters the train_and_evaluate() loop.
-                break
+                try:
+                    # Store values for the training loop.
+                    self.num_batches = wkld.num_batches
+                    self.train_response_func = response_func
+                    # Break out of the control loop so that the train process
+                    # re-enters the train_and_evaluate() loop.
+                    break
```
Comment (rb-determined-ai) [inline harness/determined/estimator/_estimator_trial.py:280]: Um, I don't think it's possible for these three lines of code to raise a `det.InvalidHP` exception.

This is the tricky part of our plugin-wrapping-a-plugin architecture; in this callback, we interrupt the main training loop to do non-training things, then our callback releases control back to the main loop, and some time later the main loop will re-enter this callback's control_loop.

That means to capture exceptions that occur during training, you'll have to do them somewhere above the scope of the estimator training loop; not inside the estimator training loop.

I actually think what @shiyuann suggested of using the push architecture for pushing checkpoints would massively simplify this.  And handing exceptions anywhere else in a way that lets you keep the harness in-tact enough to receive another message from the master asking for a checkpoint seems so hard that I wouldn't bother... I'd just say "if you raise this exception outside of evaluation you should not expect a checkpoint", and then after shiyuan's push api lands you can clean that up.

Comment (shiyuann) [reply harness/determined/estimator/_estimator_trial.py:280]: @rb-determined-ai What I have right now is the checkpoint and validation APIs, to make them work with the searcher and trial/experiment, there is still a bunch of work to do. For now, I think it's fine to just raise an exception.

Comment (rb-determined-ai) [reply harness/determined/estimator/_estimator_trial.py:280]: cool.  In that case we should probably let invalidHP exceptions during training in estimators *not* result in a checkpoint.  I think that would be too much work, and in the near future there will be an easy way to do it.

File: harness/determined/_train_context.py [hunk 17]
```diff
@@ -17,6 +17,7 @@ def __init__(self, env: det.EnvContext, hvd_config: horovod.HorovodContext):
         self.hvd_config = hvd_config  # type: horovod.HorovodContext
         self.distributed = DistributedContext(env, hvd_config)
         self._stop_requested = False
+        self._invalid_hp = False
```
Comment (rb-determined-ai) [inline harness/determined/_train_context.py:20]: is this still used anywhere?

File: harness/determined/estimator/_estimator_trial.py [hunk 18]
```diff
@@ -277,12 +278,25 @@ def control_loop(self) -> None:
...
-                response_func(
-                    det.util.wrap_metrics(
-                        self._compute_validation_metrics(),
-                        self.estimator_trial_controller.context.get_stop_requested(),
+                try:
+                    response_func(
+                        det.util.wrap_metrics(
+                            self._compute_validation_metrics(),
+                            self.estimator_trial_controller.context.get_stop_requested(),
+                            False,
```
Comment (rb-determined-ai) [inline harness/determined/estimator/_estimator_trial.py:286]: Could you use keyword argument format `metrics=False` for these?  Otherwise it is not as clear what `False` means to the reader of the code.

**Review by rb-determined-ai — COMMENTED** [review]
Last comment:

We talked not supporting checkpoint-after-InvaidHP-during-training, which is fine, but I think the way this is now written is such that we won't even tell the master there was an invalid hp at all if you raise it during training.  Is that accurate?

File: harness/determined/_train_context.py (hunk 17 above)
Comment (naren-determined) [reply harness/determined/_train_context.py:20]: Nope, good catch.

**Comment by naren-determined** [comment]
> Last comment:
> 
> We talked not supporting checkpoint-after-InvaidHP-during-training, which is fine, but I think the way this is now written is such that we won't even tell the master there was an invalid hp at all if you raise it during training. Is that accurate?

Well in `layers/_workload_manager.py` at the sections starting in Lines 146 and 203 we specifically handle how training and validation workloads respond to the master with an invalid_hp so it can then inform the searcher, right?

**Comment by rb-determined-ai** [comment]
But those process responses inside `respond_func`, and if an exception is raised in training but never caught, how will any `response_func` ever get called?

**Comment by naren-determined** [comment]
> But those process responses inside `respond_func`, and if an exception is raised in training but never caught, how will any `response_func` ever get called?

Oh sorry, yes. For PyTorch it will be caught during training, but for Estimators and Keras it will not be the way it's written at the moment. Eventually, once push architecture comes in this will be updated.

**Review by rb-determined-ai — APPROVED** [review]
ok.  I think that the way it's written now is not a good user experience, but if we are not announcing this feature until after push api lands, and we fix this issue before we do announce it, I'm fine with the current limitations.
